subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/impl/HierarchicalFileWatcherUpdater.java

```diff
@@ -64,29 +64,22 @@ public void changed(Collection<CompleteFileSystemLocationSnapshot> removedSnapsh
     }
 
     @Override
-    public void updateMustWatchDirectories(Collection<File> updatedWatchDirectories) {
-        Set<Path> rootPaths = updatedWatchDirectories.stream()
-            .filter(File::isDirectory)
-            .map(File::toPath)
-            .map(Path::toAbsolutePath)
-            .collect(Collectors.toSet());
-        Multisets.removeOccurrences(shouldWatchDirectories, mustWatchDirectories);
-        mustWatchDirectories.clear();
-        mustWatchDirectories.addAll(WatchRootUtil.resolveRootsToWatch(rootPaths));
-        shouldWatchDirectories.addAll(mustWatchDirectories);
+    public void updateProjectRootDirectory(File projectRoot) {
+        projectRootDirectory = projectRoot.toPath().toAbsolutePath();
+        projectRootDirectoryPrefix = projectRootDirectory.toString() + File.separator;
+
         updateWatchedDirectories();
     }
 
     private void updateWatchedDirectories() {
-        Set<String> mustWatchDirectoryPrefixes = ImmutableSet.copyOf(
-            mustWatchDirectories.stream()
-                .map(path -> path.toString() + File.separator)
-                ::iterator
-        );
-        Set<Path> directoriesToWatch = shouldWatchDirectories.elementSet().stream()
-            .filter(path -> !startsWithAnyPrefix(path.toString(), mustWatchDirectoryPrefixes))
-            .collect(Collectors.toSet());
-        directoriesToWatch.addAll(mustWatchDirectories);
+        Set<Path> directoriesToWatch = new HashSet<>();
+        shouldWatchDirectories.elementSet().forEach(shouldWatchDirectory -> {
+            if (projectRootDirectoryPrefix != null && shouldWatchDirectory.toString().startsWith(projectRootDirectoryPrefix)) {
+                directoriesToWatch.add(projectRootDirectory);
+            } else {
+                directoriesToWatch.add(shouldWatchDirectory);
+            }
+        });
```

Comment:
Nit: this might be simpler using a `map().collect()` call. Something like this:

```suggestion
        Set<Path> directoriesToWatch = shouldWatchDirectories.elementSet().stream()
            .map(shouldWatchDirectory -> projectRootDirectoryPrefix != null && shouldWatchDirectory.toString().startsWith(projectRootDirectoryPrefix)
                ? projectRootDirectory
                : shouldWatchDirectory)
            .collect(Collectors.toSet());
```